lib/stats/stats-query.c

```diff
@@ -218,77 +196,48 @@ _is_timestamp(gchar *counter_name)
 }
 
 void
-_sum_selected_counters(GList *counters, gpointer user_data)
+_sum_selected_counters(StatsCounterItem *counter, gpointer user_data, StatsFormatCb format_cb, gpointer result)
 {
-  GList *c;
   gpointer *args = (gpointer *) user_data;
   gint64 *sum = (gint64 *) args[1];
-  for (c = counters; c; c = c->next)
-    {
-      StatsCounterItem *counter = c->data;
-      if (!_is_timestamp(stats_counter_get_name(counter)))
-        *sum += stats_counter_get(counter);
-    }
+
+  if (!_is_timestamp(stats_counter_get_name(counter)))
+    *sum += stats_counter_get(counter);
 }
 
 gboolean
-_stats_query_get_sum(const gchar *expr, StatsSumFormatCb format_cb, gpointer result, gboolean must_reset)
+_stats_query_get_sum(const gchar *expr, StatsFormatCb format_cb, gpointer result, gboolean must_reset)
 {
   if (!expr)
     return FALSE;
 
-  gboolean found_match = FALSE;
+  const gchar *key_str = _setup_filter_expression(expr);
   gint64 sum = 0;
   gpointer args[] = {result, &sum};
+  gboolean found = _process_matching_counters(key_str, _sum_selected_counters, args, NULL, NULL, must_reset);
```

Comment:
it seems that it would be entirely possible that this module only allows the enumeration of counters matching a query and then the callback would do whatever it pleases with the result set.

This way, any form of aggregation would just be implemented by the caller and not by the query code.

Reply:
Sorry, I cannot get this
What you dexribe here is the original, old behavior that built a list then enumerated, and operated on the list items, that actually was an overhead both in storage and execution time
Any kinf of enumeration and filterings on a matching condition is now have to lock the cluster and its items during the operations, I cannot see a lock free solution here, coul you please contribute hwat possibility you can see here?

Reply:
The new implementation executes all callbacks/formatting while holding the big global stats lock.
By filtering, we could run the formatting/processor callback in a lockless manner, which would reintroduce the N*M complexity, but without the need for a really expensive lock (string formatting is expensive).

Reply:
would you like to have the separate name hash table maintenance back?

Reply:
would you like to have the separate name hash table maintenance back?

Reply:
I think the hashtable was useless, but filtering into a local data structure that is cheap to traverse and deallocate would end up being a cheaper O(N*M) traversal than a mutex-protected O(N).

@bazsi What do you think? The process callback usually contains expensive string methods, maybe it would be worth running them without holding any locks.

Reply:
Hmhm, the new Prometheus output is currently fully locked as well.

We can revisit this later in a separate PR.